bokehjs/src/coffee/models/canvas/cartesian_frame.coffee

```diff
@@ -42,24 +44,28 @@ export class CartesianFrame extends LayoutCanvas
         ranges[name] = extra_range
     return ranges
 
-  _get_scales: (scale_type_name, ranges, frame_range) ->
+  _get_scales: (scale, ranges, frame_range) ->
     scales = {}
+
     for name, range of ranges
-      if range.type == "Range1d" or range.type == "DataRange1d"
-        if scale_type_name == "log"
-          scale_type = LogScale
-        else
-          scale_type = LinearScale
-        range.scale_hint = scale_type_name
-      else if range.type == "FactorRange"
-        scale_type = CategoricalScale
-      else
-        logger.warn("unknown range type for range '#{name}': #{range}")
-        return null
-      scales[name] = new scale_type({
-        source_range: range
-        target_range: frame_range
-      })
+      if range instanceof DataRange1d or range instanceof Range1d
+        if scale not instanceof LogScale and scale not instanceof LinearScale
+          throw new Error("Range #{range.constructor.name} is incompatible is Scale #{scale.constructor.name}")
+        # special case because CategoricalScale is a subclass of LinearScale, should be removed in future
+        if scale instanceof CategoricalScale
+          throw new Error("Range #{range.constructor.name} is incompatible is Scale #{scale.constructor.name}")
+
+      if range instanceof FactorRange
+        if scale not instanceof CategoricalScale
+          throw new Error("Range #{range.constructor.name} is incompatible is Scale #{scale.constructor.name}")
+
+      if scale instanceof LogScale and range instanceof DataRange1d
+        range.scale_hint = "log"
+
+      s = scale.clone()
+      s.setv({source_range: range, target_range: frame_range})
+      s.connect_range_signals()
```

Comment:
You are introducing a memory leak here, because you're connecting new slots every time `_get_scales()` is called. Either clear old connections or make sure that you always connect the same slots. Note that you don't have to provide a cleanup procedure if you connect only once, because all slots are disconnected when model is destroyed or view removed.